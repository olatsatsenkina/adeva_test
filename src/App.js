import React from 'react';
import Note from './Note.js';
import Edit from './Edit.js';

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      mode: 'list',
      search: '',
      notes: []
    }
    this.changeMode = this.changeMode.bind(this);
    this.addNote = this.addNote.bind(this);
    this.saveNote = this.saveNote.bind(this);
    this.deleteNote = this.deleteNote.bind(this);
    this.cancelNote = this.cancelNote.bind(this);
  }

  componentWillMount() {
    let notes = this.fetchNotes();
    this.setState({notes: notes});
  }

  fetchNotes() {
    let notes = localStorage.getItem('notes');
    if (!notes)
      notes = [];
    else
      notes = notes.split(',');
    return notes;
  }

  changeMode(index, mode) {
    this.setState({mode: mode, selected: index});
  }

  addNote(text) {
    let {notes} = this.state;
    notes.push(this.replaceCommas(text));
    this.updateNotes(notes);
  }

  saveNote(text) {
    let {notes, selected} = this.state;
    notes[selected] = this.replaceCommas(text);
    this.updateNotes(notes);
  }

  deleteNote() {
    let {notes, selected} = this.state;
    notes.splice(selected, 1);
    this.updateNotes(notes);
  }

  cancelNote() {
    this.setState({mode: 'view'});
  }

  updateNotes(notes) {
    this.setState({notes: notes, mode: 'view', selected: null});
    localStorage.setItem('notes', notes);
  }

  replaceCommas(text) {
    return text.replace(/,/g,"&#44;");
  }

  searchNotes(el) {
    let searchString = el.currentTarget.value;
    this.setState({search: searchString})
  }

  render() {
    let {notes, mode, selected, search} = this.state;
    let selectedNote = notes[selected];
    let self = this;
    let noteList = notes.filter(
      x => x.indexOf(search) != -1
    ).map((el, index) =>
      (
        <Note
          handler={self.changeMode}
          text={el}
          key={"note" + index}
          index={index}/>
      )
    );
    let noNotesMessage = "No notes yet. Add some :)"
    if (notes.length) {
      noNotesMessage = "No such notes found :( "
    }
    return (
      <div className="container">
        <div className="sidebar">
          <h1>Notes</h1>
          <input
            type="search"
            value={search}
            onChange={this.searchNotes.bind(this)}
            placeholder="Search notes"
            className="search md-ish" />
          <ul>
            { noteList }
            { !noteList.length && <li className="no-notes">{noNotesMessage}</li>}
          </ul>
          <button
            className="add-button md-ish hover-cursor"
            onClick={this.changeMode.bind(this, null, 'add')}> Add Note</button>
        </div>
        { mode == 'edit' &&
          <Edit mode='edit'
            text={selectedNote}
            onSave={this.saveNote}
            onDelete={this.deleteNote}/>}
        { mode == 'add' &&
          <Edit mode='add'
            onSave={this.addNote}
            onDelete={this.cancelNote}/>}
      </div>
    );
  }
}
