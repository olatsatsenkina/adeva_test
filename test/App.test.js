import React from 'react';
import App from '../src/App.js';
import renderer from 'react-test-renderer';

it('App renders correctly', () => {
  const tree = renderer
    .create(<App/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Main workflow test', () => {
  const appComponent = renderer.create(<App/>);
  const rootInstance = appComponent.root;
  const button = rootInstance.findByType("button");
  button.props.onClick();
  const tree = appComponent.toJSON();
  expect(tree).toMatchSnapshot();
  const textarea = rootInstance.findByType("textarea");
  let e = {currentTarget: {value: "Added note text"}};
  textarea.props.onChange(e);
  const addButton = rootInstance.findByProps({className: "save-note md-ish hover-cursor"});
  addButton.props.onClick();
  const treeAfterAdd = appComponent.toJSON();
  expect(treeAfterAdd).toMatchSnapshot();
});

test('Search workflow test', () => {
  const appComponent = renderer.create(<App/>);
  const rootInstance = appComponent.root;
  const searchNotes = rootInstance.findByProps({className: "search md-ish"});
  let e = {currentTarget: {value: "Some irrelevant text here"}}
  searchNotes.props.onChange(e);
  const tree = appComponent.toJSON();
  expect(tree).toMatchSnapshot();
  e = {currentTarget: {value: "Added note"}}
  searchNotes.props.onChange(e);
  const newTree = appComponent.toJSON();
  expect(newTree).toMatchSnapshot();
});

test('Delete note test', () => {
  const appComponent = renderer.create(<App/>);
  const rootInstance = appComponent.root;
  const note = rootInstance.findByType("li");
  note.props.onClick();
  const deleteLink = rootInstance.findByProps({className: "delete-note"});
  deleteLink.props.onClick();
  const tree = appComponent.toJSON();
  expect(tree).toMatchSnapshot();
});
