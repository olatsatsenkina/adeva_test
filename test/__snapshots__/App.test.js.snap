// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="sidebar"
  >
    <h1>
      Notes
    </h1>
    <input
      className="search md-ish"
      onChange={[Function]}
      placeholder="Search notes"
      type="search"
      value=""
    />
    <ul>
      <li
        className="no-notes"
      >
        No notes yet. Add some :)
      </li>
    </ul>
    <button
      className="add-button md-ish hover-cursor"
      onClick={[Function]}
    >
       Add Note
    </button>
  </div>
</div>
`;

exports[`Delete note test 1`] = `
<div
  className="container"
>
  <div
    className="sidebar"
  >
    <h1>
      Notes
    </h1>
    <input
      className="search md-ish"
      onChange={[Function]}
      placeholder="Search notes"
      type="search"
      value=""
    />
    <ul>
      <li
        className="no-notes"
      >
        No notes yet. Add some :)
      </li>
    </ul>
    <button
      className="add-button md-ish hover-cursor"
      onClick={[Function]}
    >
       Add Note
    </button>
  </div>
</div>
`;

exports[`Main workflow test 1`] = `
<div
  className="container"
>
  <div
    className="sidebar"
  >
    <h1>
      Notes
    </h1>
    <input
      className="search md-ish"
      onChange={[Function]}
      placeholder="Search notes"
      type="search"
      value=""
    />
    <ul>
      <li
        className="no-notes"
      >
        No notes yet. Add some :)
      </li>
    </ul>
    <button
      className="add-button md-ish hover-cursor"
      onClick={[Function]}
    >
       Add Note
    </button>
  </div>
  <div
    className="dashboard"
  >
    <h2>
      Add note
    </h2>
    <div
      className="notes-contents"
    >
      <a
        className="delete-note"
        href="#"
        onClick={[Function]}
      >
        Cancel
      </a>
      <textarea
        className="note-text md-ish"
        onChange={[Function]}
        placeholder="Note's text"
      />
      <button
        className="save-note md-ish hover-cursor"
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </div>
</div>
`;

exports[`Main workflow test 2`] = `
<div
  className="container"
>
  <div
    className="sidebar"
  >
    <h1>
      Notes
    </h1>
    <input
      className="search md-ish"
      onChange={[Function]}
      placeholder="Search notes"
      type="search"
      value=""
    />
    <ul>
      <li
        onClick={[Function]}
      >
        Added note text
      </li>
    </ul>
    <button
      className="add-button md-ish hover-cursor"
      onClick={[Function]}
    >
       Add Note
    </button>
  </div>
</div>
`;

exports[`Search workflow test 1`] = `
<div
  className="container"
>
  <div
    className="sidebar"
  >
    <h1>
      Notes
    </h1>
    <input
      className="search md-ish"
      onChange={[Function]}
      placeholder="Search notes"
      type="search"
      value="Some irrelevant text here"
    />
    <ul>
      <li
        className="no-notes"
      >
        No such notes found :( 
      </li>
    </ul>
    <button
      className="add-button md-ish hover-cursor"
      onClick={[Function]}
    >
       Add Note
    </button>
  </div>
</div>
`;

exports[`Search workflow test 2`] = `
<div
  className="container"
>
  <div
    className="sidebar"
  >
    <h1>
      Notes
    </h1>
    <input
      className="search md-ish"
      onChange={[Function]}
      placeholder="Search notes"
      type="search"
      value="Added note"
    />
    <ul>
      <li
        onClick={[Function]}
      >
        Added note text
      </li>
    </ul>
    <button
      className="add-button md-ish hover-cursor"
      onClick={[Function]}
    >
       Add Note
    </button>
  </div>
</div>
`;
